aten/src/ATen/native/quantized/cpu/qnnpack/src/q8dwconv/mp8x25-sse2-per-channel.c

```diff
@@ -0,0 +1,1098 @@
+            vacc_hi, _mm_unpackhi_epi16(vprod00_odd, vprod00_even));
+
+        const __m128i vi01 = _mm_loadl_epi64((const __m128i*)i01);
+        i01 += 8;
+        const __m128i vxi01 =
+            _mm_sub_epi16(_mm_unpacklo_epi8(vi01, vzero), vinput_zero_point);
+        const __m128i vk01 =
+            _mm_loadl_epi64((const __m128i*)((uintptr_t)w + 40));
+        const __m128i vxk01 =
+            _mm_sub_epi16(
+                _mm_unpacklo_epi8(vk01, vzero),
+                _mm_unpacklo_epi8(vkernel_zero_point, vzero));
+        const __m128i vprod01_odd = _mm_mullo_epi16(vxi01, vxk01);
+        const __m128i vprod01_even = _mm_mulhi_epi16(vxi01, vxk01);
+        vacc_lo = _mm_add_epi32(
+            vacc_lo, _mm_unpacklo_epi16(vprod01_odd, vprod01_even));
+        vacc_hi = _mm_add_epi32(
+            vacc_hi, _mm_unpackhi_epi16(vprod01_odd, vprod01_even));
+
+        const __m128i vi02 = _mm_loadl_epi64((const __m128i*)i02);
+        i02 += 8;
+        const __m128i vxi02 =
+            _mm_sub_epi16(_mm_unpacklo_epi8(vi02, vzero), vinput_zero_point);
+        const __m128i vk02 =
+            _mm_loadl_epi64((const __m128i*)((uintptr_t)w + 48));
+        const __m128i vxk02 =
+            _mm_sub_epi16(
+                _mm_unpacklo_epi8(vk02, vzero),
+                _mm_unpacklo_epi8(vkernel_zero_point, vzero));
+        const __m128i vprod02_odd = _mm_mullo_epi16(vxi02, vxk02);
+        const __m128i vprod02_even = _mm_mulhi_epi16(vxi02, vxk02);
+        vacc_lo = _mm_add_epi32(
+            vacc_lo, _mm_unpacklo_epi16(vprod02_odd, vprod02_even));
+        vacc_hi = _mm_add_epi32(
+            vacc_hi, _mm_unpackhi_epi16(vprod02_odd, vprod02_even));
+
+        const __m128i vi10 = _mm_loadl_epi64((const __m128i*)i10);
+        i10 += 8;
+        const __m128i vxi10 =
+            _mm_sub_epi16(_mm_unpacklo_epi8(vi10, vzero), vinput_zero_point);
+        const __m128i vk10 =
+            _mm_loadl_epi64((const __m128i*)((uintptr_t)w + 56));
+        const __m128i vxk10 =
+            _mm_sub_epi16(
+                _mm_unpacklo_epi8(vk10, vzero),
+                _mm_unpacklo_epi8(vkernel_zero_point, vzero));
+        const __m128i vprod10_odd = _mm_mullo_epi16(vxi10, vxk10);
+        const __m128i vprod10_even = _mm_mulhi_epi16(vxi10, vxk10);
+        vacc_lo = _mm_add_epi32(
+            vacc_lo, _mm_unpacklo_epi16(vprod10_odd, vprod10_even));
+        vacc_hi = _mm_add_epi32(
+            vacc_hi, _mm_unpackhi_epi16(vprod10_odd, vprod10_even));
+
+        const __m128i vi11 = _mm_loadl_epi64((const __m128i*)i11);
+        i11 += 8;
+        const __m128i vxi11 =
+            _mm_sub_epi16(_mm_unpacklo_epi8(vi11, vzero), vinput_zero_point);
+        const __m128i vk11 =
+            _mm_loadl_epi64((const __m128i*)((uintptr_t)w + 64));
+        const __m128i vxk11 =
+            _mm_sub_epi16(
+                _mm_unpacklo_epi8(vk11, vzero),
+                _mm_unpacklo_epi8(vkernel_zero_point, vzero));
```

Comment:
The ```_mm_unpacklo_epi8(vkernel_zero_point, vzero)```s are repeated several times below it seems.  Does anything change between each invocation?  If not, you can consider factoring them out to the outer most loop at a point where the inputs to _mm_unpacklo_epi8 first become available.

Reply:
They are not loop invariant since `verkenel_zero_point` is loaded every iteration, but yes within the same loop they dont need to be repeated. I am assuming that compiler can atleast do some common subexpression elimination. But more honestly since this is just sse2 code, I did not worry whether compiler actually does this or not.